table.go

```diff
@@ -1,75 +1,181 @@
-			}
+	default: // "string"
+		if len(q) == 0 {
+			return nil, nil
 		}
+		return string(q[1:]), nil
+	}
+}
+
+type cacheKey [][]byte
+
+func decodePartitionKey(t r.Table, fvn r.FieldValuesByName) error {
+	for _, fn := range t.PartitionKey {
+		fv, ok := fvn[fn]
+		if !ok {
+			return MissingFieldValueError{t.Name, fn, fvn}
+		}
+		if _, err := marshalField(t, fn, fv); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func decodePrimaryKey(t r.Table, fvn r.FieldValuesByName) (cacheKey, error) {
+	ck := make(cacheKey, 0, len(t.PrimaryKeyCachePrefix)+len(t.PrimaryKey))
+	for _, k := range t.PrimaryKeyCachePrefix {
+		ck = append(ck, []byte(k))
 	}
 	for _, fn := range t.PrimaryKey {
-		ft, ok := t.Types[fn]
+		fv, ok := fvn[fn]
 		if !ok {
-			return fmt.Errorf("Table %v: Missing FieldType for FieldName: %v", t.Name, fn)
+			return nil, MissingFieldValueError{t.Name, fn, fvn}
 		}
-		if ft != "string" {
-			return fmt.Errorf("Table %v: Invalid FieldType for FieldName %v: %v", t.Name, fn, ft)
+		v, err := marshalField(t, fn, fv)
+		if err != nil {
+			return nil, err
 		}
+		ck = append(ck, v)
 	}
+	return ck, nil
+}
+
+func (k cacheKey) key() c.CacheKey {
+	b := bytes.Buffer{}
+	for i, q := range k {
+		if i != 0 {
+			// TODO use interface
+			b.WriteRune('#')
+		}
+		b.Write(q)
+	}
+	return c.CacheKey(b.Bytes())
+}
+
+type cacheValue [][]byte
+
+func decodeValue(t r.Table, fvn r.FieldValuesByName) (cacheValue, error) {
+	cv := make(cacheValue, 0, len(t.Values))
 	for _, fn := range t.Values {
-		ft, ok := t.Types[fn]
+		fv, ok := fvn[fn]
 		if !ok {
-			return fmt.Errorf("Table %v: Missing FieldType for FieldName: %v", t.Name, fn)
+			cv = append(cv, nil)
+			continue
 		}
-		if ft != "string" {
-			return fmt.Errorf("Table %v: Invalid FieldType for FieldName %v: %v", t.Name, fn, ft)
+		v, err := marshalField(t, fn, fv)
+		if err != nil {
+			return nil, err
 		}
+		cv = append(cv, v)
 	}
-	return nil
+	return cv, nil
+}
+
+func (v cacheValue) value() c.CacheValue {
+	if v == nil {
+		return nil
+	}
+	b := bytes.Buffer{}
+	for i, q := range v {
+		if i != 0 {
+			// TODO use interface
+			b.WriteRune('#')
+		}
+		b.Write(q)
+	}
+	cv := c.CacheValue(b.Bytes())
+	if cv == nil {
+		cv = c.CacheValue{}
+	}
+	return cv
+}
+
+func split(cv c.CacheValue) (v cacheValue) {
+	if cv == nil {
+		return nil
+	}
+	return cacheValue(bytes.Split(cv, []byte("#")))
+}
+
+func isMissingSomeValue(t r.Table, fvn r.FieldValuesByName) bool {
+	for _, fn := range t.Values {
+		if _, ok := fvn[fn]; !ok {
+			return true
+		}
+	}
+	return false
+}
+
+func encode(t r.Table, kv r.FieldValuesByName, cv cacheValue) (r.FieldValuesByName, error) {
+	if cv == nil {
+		return nil, nil
+	}
+	fvn := make(r.FieldValuesByName, len(t.PrimaryKey)+len(t.Values))
+	for _, fn := range t.PrimaryKey {
+		fvn[fn] = kv[fn]
+	}
+	for i, fn := range t.Values {
+		if i >= len(cv) {
+			break
+		}
+		v, err := unmarshalField(t, fn, cv[i])
+		if err != nil {
```

Comment:
Maybe "else if" would be more efficient here

Reply:
Maybe, but I don't see it like a choice between two alternatives, but two diffferent checks that must be done.